src/EditorFeatures/CSharpTest/Diagnostics/GenerateMethod/GenerateDeconstructMethodTests.cs

```diff
@@ -0,0 +1,300 @@
+}",
+@"using System;
+
+class Class
+{
+    void Method()
+    {
+        (int x, int y) = this;
+    }
+    void Deconstruct(out int x) => throw null;
+    void Deconstruct(out int x, out int y, out int z) => throw null;
+
+    private void Deconstruct(out int x, out int y)
+    {
+        throw new NotImplementedException();
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionDeclaration_AlreadySuccessfull()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        (int x, int y) = [|this|];
+    }
+    void Deconstruct(out int x, out int y) => throw null;
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionDeclaration_UndeterminedType()
+        {
+            await TestInRegularAndScript1Async(
+@"class Class
+{
+    void Method()
+    {
+        (var x, var y) = [|this|];
+    }
+}",
+@"using System;
+
+class Class
+{
+    private void Deconstruct(out object x, out object y)
+    {
+        throw new NotImplementedException();
+    }
+
+    void Method()
+    {
+        (var x, var y) = this;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionDeclaration_UndeterminedType2()
+        {
+            await TestInRegularAndScript1Async(
+@"class Class
+{
+    void Method()
+    {
+        var (x, y) = [|this|];
+    }
+}",
+@"using System;
+
+class Class
+{
+    private void Deconstruct(out object x, out object y)
+    {
+        throw new NotImplementedException();
+    }
+
+    void Method()
+    {
+        var (x, y) = this;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionDeclaration_BuiltinType()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        (int x, int y) = [|1|];
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionAssignment()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        int x, y;
+        (x, y) = [|this|];
+    }
+}",
+@"using System;
+
+class Class
+{
+    private void Deconstruct(out int x, out int y)
+    {
+        throw new NotImplementedException();
+    }
+
+    void Method()
+    {
+        int x, y;
+        (x, y) = this;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionAssignment_Nested()
+        {
+            // We only offer a fix for non-nested deconstruction, at the moment
```

Comment:
>// We only offer a [](start = 12, length = 18)

Should there be an issue tracking this?

Reply:
Nested deconstructions aren't very common, so I'd rather wait for someone to raise the question. Otherwise, the issue is likely to just sit there forever.